src/clients/c/tb_client/context.zig

```diff
@@ -308,10 +332,19 @@ pub fn ContextType(
                 return self.on_complete(packet, error.InvalidDataSize);
             }
 
-            // Make sure the packet.data wouldn't overflow a message:
-            if (events.len > constants.message_body_size_max) {
+            // Make sure the packet.data wouldn't overflow a request, and that the corresponding
+            // results won't overflow a reply.
+            const events_batch_max = switch (operation) {
+                .pulse => unreachable,
+                inline else => |operation_comptime| StateMachine.operation_batch_max(
+                    operation_comptime,
+                    self.batch_size_limit.?,
+                ),
+            };
+            if (@divExact(events.len, event_size) > events_batch_max) {
                 return self.on_complete(packet, error.TooMuchData);
```

Comment:
This is unrelated to the PR, but I am wondering, with auto-batching, do we have a TOCTOU here, where we check size in `request`, but use later, after some user's code has been run, in `submit`?

We are definitely missing assertions there!

Reply:
cc @batiati